src/main.rs

```diff
@@ -1,47 +1,216 @@
+#![allow(clippy::unusual_byte_groupings)]
+#![allow(clippy::upper_case_acronyms)]
 use std::fs;
 
 const MEMORY_SIZE: usize = 2_usize.pow(16);
 
+#[derive(Debug)]
+enum Argument {
+    Reg(u16),
+    Immediate(i16),
+}
+
 #[derive(Debug)]
 enum Opcode {
-    BR, // Branch
-    ADD, // Add
-    LD, // Load
-    ST, // Store
-    JSR, // Jump Register
-    AND, // Bitwise And
-    LDR, // Load Register
-    STR, // Store Register
-    RTI, // Unused
-    NOT, // Bitwise Not
-    RET, // Return from JSR
-    LDI, // Load Indirect
-    STI, // Store Indirect
-    JMP, // Jump
-    RES, // Unused
-    LEA, // Load Effective Address
-    TRAP // Execute Trap
+    ADD {
+        dr: u16,
+        sr1: u16,
+        sr2: Argument,
+    }, // Add (dr <- sr1 + sr2/imm5)
```

Comment:
Very cool use of Rust enums with parameters!  Love it!

Very minor nit-pick, could the comments be written next to the enum name? Like so:

```rust
    ADD { // Add (dr <- sr1 + sr2/imm5)
        dr: u16,
        sr1: u16,
        sr2: Argument,
    }, 
```

Instead of 

```rust
    ADD {
        dr: u16,
        sr1: u16,
        sr2: Argument,
    }, // Add (dr <- sr1 + sr2/imm5)
```

I think it is slightly clearer. What do you think?

Reply:
I agree but `cargo fmt` doesnt allow it, it has to be either that or:

```rust
ADD {
        // Add (dr <- sr1 + sr2/imm5)
        dr: u16,
        sr1: u16,
        sr2: Argument,
    },
```

Which adds 1 extra line per comment, what do you think is better?

Reply:
> I agree but `cargo fmt` doesnt allow it, it has to be either that or:

Darn it.

 
> ```rust
> ADD {
>         // Add (dr <- sr1 + sr2/imm5)
>         dr: u16,
>         sr1: u16,
>         sr2: Argument,
>     },
> ```

It think I like this slightly better. Although, I have to admit it's pretty subjective.

Reply:
Ok I'll change that